babel/messages/extract.py

```diff
@@ -128,6 +129,9 @@ def extract_from_dir(dirname=None, method_map=DEFAULT_MAPPING,
                      positional arguments, in that order
     :param strip_comment_tags: a flag that if set to `True` causes all comment
                                tags to be removed from the collected comments.
+    :param directory_filter: a list of strings that identify the starting
```

Comment:
To be honest, this might be more useful if it declared a list of [`fnmatch`](https://docs.python.org/3.5/library/fnmatch.html#module-fnmatch) compatible patterns.

Also, now that I think of it, it could maybe just accept a callable? After all there's no command-line support for this argument yet, so it needn't be a "command line compatible". The default could be `default_directory_filter`,  defined as 

```python
def default_directory_filter(path):
  subdir = os.path.basename(path)
  return not (subdir.startswith('.') or subdir.startswith('_'))
```

and the implementation in the walk loop simply
```python
dirnames[:] = [dirname for dirname in dirnames if directory_filter(os.path.join(root, dirname)]
```